File: app/api_client.rb
class APIClient
  def self.get_etas(station_id)
    url = 'http://lapi.transitchicago.com/api/1.0/ttarrivals.aspx?key=%s&mapid=%d' % [Secrets.api_key, station_id]
    error_ptr = Pointer.new(:object)
    data = NSData.alloc.initWithContentsOfURL(NSURL.URLWithString(url), options: NSDataReadingUncached, error: error_ptr)

    return nil if !data

    delegate = ArrivalXMLParserDelegate.new
    parser = NSXMLParser.alloc.initWithData(data)
    parser.delegate = delegate
    parser.parse

    delegate.etas.sort
  end

  def self.get_alerts
    url = 'http://www.transitchicago.com/api/1.0/alerts.aspx'
    error_ptr = Pointer.new(:object)
    data = NSData.alloc.initWithContentsOfURL(NSURL.URLWithString(url), options: NSDataReadingUncached, error: error_ptr)

    return nil if !data

    delegate = AlertXMLParserDelegate.new
    parser = NSXMLParser.alloc.initWithData(data)
    parser.delegate = delegate
    parser.parse

    delegate.alerts.select { |alert| alert.affects_trains? }.uniq { |alert| alert.description }
  end
end

File: Rail Line/APIClient.swift
import Foundation

class APIClient {
    static func getETAs(stationId: Int) -> [ETA] {
        let url = "http://lapi.transitchicago.com/api/1.0/ttarrivals.aspx?key=\(Secrets.apiKey)&mapid=\(stationId)"
        let delegate = ArrivalXMLParserDelegate()
        let parser = XMLParser(contentsOf: URL(string: url)!)
        parser!.delegate = delegate
        parser!.parse()

        return delegate.etas.sorted(by: { $0 < $1 })
    }

    static func getAlerts() -> [Alert] {
        let url = "http://www.transitchicago.com/api/1.0/alerts.aspx"
        let delegate = AlertsXMLParserDelegate()
        let parser = XMLParser(contentsOf: URL(string: url)!)
        parser!.delegate = delegate
        parser!.parse()

        // only show alerts that affect trains and don't show ones with identical descriptions
        var descriptions: Set<String> = Set()
        return delegate.alerts.filter { $0.affectsTrains() }.filter {
            if descriptions.contains($0.description) { return false }
            descriptions.insert($0.description)
            return true
        }
    }
}
